drivers: display: gc9x01x: Add support for galaxy core gc9a01a display controller

MrMarteng opened the pull request:
Adding display driver for GC9A01A LCD display controller from galaxycore.

amrithvenkat commented:
Hi @MrMarteng Thanks for taking up the changes to the driver. There are some compliance issues that need to be resolved. Please look into them. :)

Y009 commented:
Might be worth renaming it to gc9a01a -> gc9x01x
as I was looking at gc9c01 and it seems to be almost entirely the same, with the same commands
The c version is bigger, with a 360x360 resolution and ~2x more GRAM.
for ref: [https://www.buydisplay.com/download/ic/GC9C01.pdf](https://www.buydisplay.com/download/ic/GC9C01.pdf)

Commit:
drivers: display: gc9a01a: Add support for SPI display controller gc9a01a

Adding driver for GC9A01A 240x240 based LCD displays.
Should be working with GC9C01 as well (untested).

Signed-off-by: Martin Kiepfer <mrmarteng@teleschirm.org>

MrMarteng commented:
Still work in progress. I am also still trying to get details on some undocumented registers/commands.

ljunquera commented:
Thanks, @MrMarteng. Do you have an overlay you were testing with?

amrithvenkat commented on include/zephyr/dt-bindings/display/gc9x01x.h:
```diff
@@ -0,0 +1,19 @@
+/* Display rotation */
+#define GC9X01X_ORIENTATION_NORMAL 0U
+#define GC9X01X_ORIENTATION_90     90U
+#define GC9X01X_ORIENTATION_180    90U
```
```suggestion
#define GC9X01X_ORIENTATION_180    180U
```

MrMarteng replied:
Fixed

MrMarteng commented:
> Thanks, @MrMarteng. Do you have an overlay you were testing with? I'm going to try it with the lvgl base sample, [Waveshare](https://www.waveshare.com/wiki/1.28inch_Touch_LCD#Support), and a nRF5340 DK:
> 
> ```
>     mydisplay: gc9x01x@0 {
>         compatible = "galaxycore,gc9x01x";
>         reg = <0>;
>         reset-gpios = <&gpio1 10 GPIO_ACTIVE_HIGH>;
>         cmd-data-gpios = <&gpio1 11 GPIO_ACTIVE_HIGH>;
>         backlight-pwm = <&pwm_led0>;
>         pixel-format = <0>;
>         display-inversion;
>         spi-max-frequency = <8000000>;
>         width = <240>;
>         height = <240>;
>     };
> ```
> 
> Has anybody else tried to get it working yet?

I am testing it on a m5dial. Here is my working configuration:
```
/ {
    pwmleds {
      compatible = "pwm-leds";
      backlight_pwm: backlight_pwm {
        pwms = <&ledc0 0 PWM_MSEC(10) PWM_POLARITY_NORMAL>;
      };
    };
  };

  &spi2 {
    gc9a01a_lcd: gc9a01a_lcd@0 {
      compatible = "galaxycore,gc9x01x";
      reg = <0>;
      spi-max-frequency = <100000000>;
      cmd-data-gpios = <&gpio0 4 GPIO_ACTIVE_HIGH>;
      reset-gpios = <&gpio0 8 GPIO_ACTIVE_HIGH>;
      backlight-pwm = <&backlight_pwm>;

      width = <240>;
      height = <240>;
      pixel-format = <GC9X01X_PIXEL_FORMAT_RGB565>;
    };
  };
```

ljunquera commented:
...
> I also added an example to the driver definition file.

Thank you for all the help. I'm struggling a little to get the hello world running. I'll work through the issues as they are probably on my end. I have it working with an st7789.

MrMarteng commented:
> ...
> 
> > I also added an example to the driver definition file.
> 
> Thank you for all the help. I'm struggling a little to get the hello world running. I'll work through the issues as they are probably on my end. I have it working with an st7789.



> ...
> 
> > I also added an example to the driver definition file.
> 
> Thank you for all the help. I'm struggling a little to get the hello world running. I'll work through the issues as they are probably on my end. I have it working with an st7789.

Can you describe what problem you face?
I would recommend to move this discussion somewhere else to keep this PR clean. Feel free to write me a PM on discord.

amrithvenkat commented:
I think the vendor prefixes need to be added in `/dts/bindings/vendor-prefixes.txt`

`node '/soc/spi@4003c000/mydisplay@0' compatible 'galaxycore,gc9x01x' has unknown vendor prefix 'galaxycore'` - during compilation.

MrMarteng commented:
> I think the vendor prefixes need to be added in `/dts/bindings/vendor-prefixes.txt`
> 
> `node '/soc/spi@4003c000/mydisplay@0' compatible 'galaxycore,gc9x01x' has unknown vendor prefix 'galaxycore'` - during compilation.

Thanks. That change has been accidentally moved to another branch.
That brings me to the question if there is any tests the display driver needs to be added to?

amrithvenkat commented:
As far as i know you can add a test to `/tests/drivers/build-all/display/app.overlay` and add a display_gc9x01x.conf.

kartben commented:
It would be nice, and probably relatively easy AFAICT, to implement device power management? See e.g. https://github.com/zephyrproject-rtos/zephyr/blob/aa4416624dceef29ec9ca78820c9fe5d3868f62c/drivers/display/display_st7789v.c#L364

MrMarteng commented:
> It would be nice, and probably relatively easy AFAICT, to implement device power management? See e.g.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/aa4416624dceef29ec9ca78820c9fe5d3868f62c/drivers/display/display_st7789v.c#L364

Should not be a problem. How can I actually test the if it is working?
I couldn't find a PM shell and I am generally unsure if the use-count of the display will ever be 0?

MrMarteng commented:
> > It would be nice, and probably relatively easy AFAICT, to implement device power management? See e.g.
> 
> > 
> 
> > https://github.com/zephyrproject-rtos/zephyr/blob/aa4416624dceef29ec9ca78820c9fe5d3868f62c/drivers/display/display_st7789v.c#L364
> 
> 
> 
> Should not be a problem. How can I actually test the if it is working?
> 
> I couldn't find a PM shell and I am generally unsure if the use-count of the display will ever be 0?
> 
> 
> 

I found an example and reused it for testing. PM support has been tested and changes are added to this PR.

MrMarteng commented:
Any additional comments, feedback?

akscram commented on drivers/display/Kconfig.gc9x01x:
```diff
@@ -0,0 +1,11 @@
+	default y
+	depends on DT_HAS_GALAXYCORE_GC9X01X_ENABLED
+	select SPI
+	select PWM
```
Is `select GPIO` missing here?

MrMarteng replied:
Generally you are right, but I don't see it on the other SPI based drivers as well (ili, st). And they all utilize GPIOs as well.

akscram replied:
So, what would be the right approach then?

akscram commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+#define GC9X01X_GPIO_LEVEL_DATA 1U
+
+/* Maximum number of default init registers  */
+#define GC9X01X_NUM_DEFAULT_INIT_REGS (16U)
```
According to commands listed below, the maximum number is 12. Would it be better to lower this value from 16 to 12?

MrMarteng replied:
I left some buffer, but I agree that it is better to keep things tight!

akscram commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+		.data = {0x28U, 0x29U, 0xF1U, 0x01U, 0xF1U, 0x00U, 0x07U},
+	},
+	{
+		.cmd = 0x66U,
```
I found out that only 0x66 and 0x67 commands can't be removed and 0x84 mirrors X but it can be compensated by picking parameters for 0x36 which will also require to figure out how to properly set the requested orientation. The other undocumented commands, when removed, don't seem to affect rendering, at least visually.

akscram commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+
+	if (desc->pitch > desc->width) {
+		write_h = 1U;
+		nbr_of_writes = desc->height;
```
I'm wondering when the pitch can be greater than width? And why are we switching in line by line writes in this case?

MrMarteng replied:
From my understanding pitch can be used to skip areas in the provided buffer (e.g. for interlacing or just copying a subpart of a loaded resource. In this case pitch can be larger than width. The remaining bytes in a pixel column therefore is calculated based on pitch instead of width.

akscram commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int gc9x01x_pm_action(const struct device *dev, enum pm_device_action action)
```
I'm wondering whether PWM, GPIO and API also should be resumed and suspended.

MrMarteng replied:
I am not an PM expert, but from what I understand in terms of device power management, every device needs to keep track of its used count (`pm_device_runtime_get`, `pm_device_runtime_put`). For the display driver the common layer will take care of this and this driver only implements/provides the actual execution of the according power management mode. Am I wrong, or what specifically do you want me implement here?

akscram replied:
I'm not an expert in this too. I thought that as these peripherals, PWM, GPIO and SPI, are initialized and controlled in the driver, the driver should be responsible for their PM but I might be wrong. It's better to do some research or ask someone who is familiar with this.

MrMarteng replied:
I am unsure, what you want me to do here. I tested behaviour with a small sample which worked without problems.

As the display driver is not part of a power domain it can be suspended and resumed using the [device power management API](https://docs.zephyrproject.org/latest/services/pm/device.html#) directly.  Where do you see an issue here?

akscram commented on drivers/display/display_gc9x01x.h:
```diff
@@ -0,0 +1,101 @@
+/* Command registers */
+#define GC9X01X_CMD_SLPIN     (0x10U) /* Enter Sleep Mode */
+#define GC9X01X_CMD_SLPOUT    (0x11U) /* Exit Sleep Mode */
+#define GC9X01X_CMD_PTLON     (0x12U) /* Partial Mode ON */
```
Some of macros are not used. Are they for documentation purpose here?

MrMarteng replied:
Yes, it's for documentation and completeness purpose. I added definitions for all registers from the datasheet. If someone else is going to extend functionality he can make use of this. These defines do not consume any memory, so I think that should be fine.

akscram commented on dts/bindings/display/galaxycore,gc9x01x.yaml:
```diff
@@ -0,0 +1,169 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  GC9X01X display controller.
```
Should any specific modules, tested, be listed here? It would make this driver much better searchable.

MrMarteng replied:
Good point. I updated the yaml description to make this more clear.
Fixed with [8052345](https://github.com/zephyrproject-rtos/zephyr/commit/805234525391bf58e9868afa0ab5b0395bda12fe)

akscram commented on dts/bindings/display/galaxycore,gc9x01x.yaml:
```diff
@@ -0,0 +1,169 @@
+    type: int
+    default: 0
+    enum:
+      - 0 # RGB565
```
Would it be better to use string enums, e.g. "rgb565" and "rgb888", which are much more selfdescriptive?

MrMarteng replied:
There actually exist definitions in `gc9x91x.h` file. I updated the yaml description to make this more clear.

MrMarteng replied:
Fixed with latest push

joelguittet commented:
@MrMarteng 
I finally get my hardware running (ESP32-S3-Touch-LCD-1.28 from Waveshare - I will submit a PR for this board if your driver is being merged).
Just one question please while using the `samples/drivers/display` application, the 4 rectangles on the screen should be RED, GREEN, BLUE and GREY. Is it correct on your hardware ?
On my board all bits are reversed i.e . to display RED: 0x00FFFF, GREEN: 0xFF00FF, BLUE: 0xFFFF00. Maybe we have not the same display (even if it is the same controller).
Except this comment, nothing to say, it's working like a charm.
Thanks,
Joel

joelguittet commented:
@MrMarteng 
Got it, I just added `display-inversion;` in my device tree, thanks for supporting this on this driver as well :-)
This is all fine for me, working as expected.

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+#define GC9X01X_GPIO_LEVEL_DATA 1U
+
+/* Maximum number of default init registers  */
+#define GC9X01X_NUM_DEFAULT_INIT_REGS (12U)
```
nit: no need for the parenthesis around this one

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+	uint16_t y_resolution;
+	bool inversion;
+	const void *regs;
+	int (*regs_init_fn)(const struct device *dev);
```
is this needed? future proofing? If there's no plan to use it just drop it, can always add it back later

MrMarteng replied:
That was left over from a previous version.
Fixed in. [3af98ab](https://github.com/zephyrproject-rtos/zephyr/commit/3af98ab51e4a42df60a3209ef92480cd397189a3)

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+
+	if (orientation == DISPLAY_ORIENTATION_NORMAL) {
+		/* works 0° */
+		tx_data |= 0;
```
maybe drop it and add a comment that that's the default setting?

MrMarteng replied:
[3af98ab](https://github.com/zephyrproject-rtos/zephyr/commit/3af98ab51e4a42df60a3209ef92480cd397189a3)

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+{
+	const struct gc9x01x_config *config = dev->config;
+
+	if (config->backlight.dev != NULL) {
+		uint32_t step = config->backlight.period / 100U;
+		int retval = pwm_set_pulse_dt(&config->backlight, brightness * step / 2.55);
+
+		if (retval) {
+			LOG_ERR("Failed to set pulse width\n");
+			return retval;
+		}
+		return 0;
+	}
```
This is normally implemented only if the display controller itself includes backlight control, so I think this should just return -NOTSUP in this case.

I realize this is needed for the samples to work though so maybe we should have a known node and have the samples be aware of that, regardless, that should be fixed separately from this PR.

MrMarteng replied:
I am aligned that this is not the right solution, but for the board I am currently working with, backlight is only working when the PWM is initialized. All sample applications would show a dark screen as there is currently no possibility to set an initial PWM in device tree.

IMO a pwm backlight driver would be helpful.(https://github.com/torvalds/linux/blob/9f8413c4a66f2fb776d3dc3c9ed20bf435eb305e/drivers/video/backlight/pwm_bl.c)
Maybe we could make use of the `led_driver_api`?

fabiobaltieri replied:
Yeah I think this should use the LED API, how about:
- drop the pwm stuff from here
- define the backlight pwm in the board file as a `pwm-leds` node, for example https://github.com/zephyrproject-rtos/zephyr/pull/67311/files for example:
```
	pwmleds {
		compatible = "pwm-leds";
		backlight_pwm: backlight_pwm {
			pwms = <&ledc0 0 PWM_HZ(250) PWM_POLARITY_NORMAL>;
		};
	};
```
- link the node to a well known chosen, maybe `zephyr,display-backlight`
```

	chosen {
		zephyr,display = &gc9a01;
		zephyr,display-backlight = &backlight_pwm;
	};
```
- set `CONFIG_LED=y` in the LVGL samples
- in the LVGL samples check if the backlight chosen node is defined, if it is find the device with `DEVICE_DT_GET(DT_PARENT(DT_CHOSEN(...)))`, the node offset with `DT_NODE_CHILD_IDX(DT_CHOSEN(...))` and then do a `led_set_brightness(dev, idx, value);` in the application

@faxe1008 @jfischer-no would that make sense?

Linux has a dedicated driver because there's some userspace APIs specifically for backlight but for Zephyr things are a bit simpler, I think we can get away with the existing LED APIs.

MrMarteng replied:
I have a similar definition in the board layer already, but **I think the missing piece is the initialisation an initial PWM value via devicetree**. It should be possible on board level to set an adequate PWM init value. If it is the application being responsible we would need to rework all existing display related sample applications.

fabiobaltieri replied:
yeah I'm suggesting to rework the samples (the last point I made)

MrMarteng replied:
> Linux has a dedicated driver because there's some userspace APIs specifically for backlight but for Zephyr things are a bit simpler, I think we can get away with the existing LED APIs.
We don't need a full api for that. The `led_driver_api` already features all we need, including `led_set_brightness`.

What do you think about the possibility to extend the `pwm-leds` driver with an init value?

For better clarity it may also a possibility to create more specific variant of the `pwm-leds` driver (e.g. `pwm-backlight`) that features even more functionality related to display backlight control. But I don't think that's needed for now.

MrMarteng replied:
@faxe1008 @jfischer-no What's your opinion on that:
- Do we prefer to rework the graphics examples to set a default brightness when `zephyr,display-backlight` is specified? Are there additional tasks needed for this solution (e.g.documentation)?
- Or do you think it makes sense to fruther think about a possibility to set a default PWM value in the devicetreee for that boards that only rely on PWM-controller backlight?

faxe1008 replied:
Since there are cases where the backlight is a device config (settings or similar), having it as a DT default value does not feel like an elegant solution. I get that it makes it more user friendly because there is one less thing to think of. I'd love to be this part of the display API, but for simplicity sake the using the led api is ok imo.

MrMarteng replied:
As per request I removed the embedded backlight control from the driver ([3c8529d](https://github.com/zephyrproject-rtos/zephyr/pull/67012/commits/3c8529d458c1c3d833a1035cf4963e08dd0facb1)).

However, I would like to point out again that most devices with this display controller have PWM backlighting. This must be set by the application, as there is no option in existing `led_pwm` drivers to set an initial duty_cycle in the devicetree. This means that the existing graphic examples do currently not work by default in this case and the display will stay dark.

I will start working on extending the existing `led_pwm` driver or creating a new backlight driver (compared to linux) to get a working solution for the board I am developing with.

fabiobaltieri replied:
> This means that the existing graphic examples do currently not work by default

Yes that needs to be fixed would the proposal I mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/67012#discussion_r1446567202 work? Not sure what I'm asking for is clear but I'm happty to create a PR for that myself if that can help clarifying what I mean.

MrMarteng replied:
This proposal will not work (at least for my board) as there is no possibility to initialize the PWM with a dutycycle. Basically I am already using the same device tree definition in my board/shield (see #67253).

I already started working on a separate `backlight_pwm` driver that is very similar to `led_pwm`. I am still considering if reusing the led_driver_api is the right way, but I'll keep you posted.

fabiobaltieri replied:
What's wrong with initializing the level in the application?

MrMarteng replied:
For me personally it's mainly inconvenience and I could imagine that this could be a trap-fall for some people.

fabiobaltieri replied:
You could say the same for `display_blanking_off()`, if you draw your graphics and forget to unblank you also get a blank screen.

MrMarteng replied:
Hmm, good point. I will rethink about that. Maybe there are additional advantages in terms of power management having the display driver and some sort of backlight control interoperate.

fabiobaltieri replied:
My main concern with auto initializing a backlight at driver init level is that that's certainly going to happen before any application code kicked in and draw anything useful on the screen, and potentially (since pwm and display are independent devices) can happen before the screen has ever been initialized, potentially flashing garbage on the screen during the display initialization sequence. That's quite uncommon for memory based display (like this one) but I see it happening all the time with memoryless displays, the fix for that is to turn on the backlight at the right point in time (which I can't imagine not being in the application code) and I say that it'd be nice to have that in our samples anyway.

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+		return retval;
+	}
+
+	retval = gc9x01x_set_brightness(dev, 50);
+	if (retval < 0) {
+		LOG_ERR("Could not set brightness (%d)", retval);
+		return retval;
+	}
```
it should be up to the application to enable the backlight at the right moment, see my comment above

MrMarteng replied:
It may stay a proprietary solution until a pwm backlight init driver is available (See above).
For better clarity I added a separate devicetree priority to define the initial value.

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+#define DT_INST_FOREACH_GC9X01X_STATUS_OKAY                                                        \
+	LISTIFY(DT_NUM_INST_STATUS_OKAY(galaxycore_gc9x01x), GC9X01X_INIT, (;))
+
+#ifdef CONFIG_GC9X01X
```
this cannot not happen, the driver depends on `CONFIG_GC9X01X` to be built

MrMarteng replied:
fixed in [3af98ab](https://github.com/zephyrproject-rtos/zephyr/commit/3af98ab51e4a42df60a3209ef92480cd397189a3)

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,745 @@
+			 &gc9x01x_data_##n, &gc9x01x_config_##n, POST_KERNEL,                      \
+			 CONFIG_DISPLAY_INIT_PRIORITY, &gc9x01x_api)
+
+#define DT_INST_FOREACH_GC9X01X_STATUS_OKAY                                                        \
+	LISTIFY(DT_NUM_INST_STATUS_OKAY(galaxycore_gc9x01x), GC9X01X_INIT, (;))
```
just use `DT_INST_FOREACH_STATUS_OKAY`, the sample you copied from has some abstraction that I don't think it's needed here

MrMarteng replied:
init structure has been reworked in [3af98ab](https://github.com/zephyrproject-rtos/zephyr/commit/3af98ab51e4a42df60a3209ef92480cd397189a3)

MrMarteng commented on drivers/display/CMakeLists.txt:
```diff
@@ -21,6 +21,7 @@ zephyr_library_sources_ifdef(CONFIG_ST7789V		display_st7789v.c)
 zephyr_library_sources_ifdef(CONFIG_ST7735R		display_st7735r.c)
 zephyr_library_sources_ifdef(CONFIG_STM32_LTDC		display_stm32_ltdc.c)
 zephyr_library_sources_ifdef(CONFIG_RM68200		display_rm68200.c)
+zephyr_library_sources_ifdef(CONFIG_GC9X01X	display_gc9x01x.c)
```
fixed in [3af98ab](https://github.com/zephyrproject-rtos/zephyr/commit/3af98ab51e4a42df60a3209ef92480cd397189a3)

akscram reviewed (commented):
The driver implementation looks good to me besides a couple of minor moments, e.g. the default backlight value, and the fact that it uses an undocumented sequence of commands without which a controller can't be properly initialized. Thank you for implementing it! The question whether it's okay to include this undocumented sequence in the driver code and whether it could cause licensing problems still remains open from my understanding because I'm not an expert in this area and it's better to rely on opinions of other folks who are familiar with this subject.

fabiobaltieri commented:
Can you squash the two commits and push again? does not make much sense to add something just to remove it the commit after.

MrMarteng commented:
> Can you squash the two commits and push again? does not make much sense to add something just to remove it the commit after.

Commits have been combined in [03357fe](https://github.com/zephyrproject-rtos/zephyr/commit/03357fec2e38eb88eb3dd36286a633ace9b19a3a)

jfischer-no reviewed (changes requested):
A few little things, the backlight approach looked good to me. Not sure why it was removed.

MrMarteng commented on dts/bindings/display/galaxycore,gc9x01x.yaml:
```diff
@@ -0,0 +1,181 @@
+      high(1) for data, low(0) for command.
+
+  pixel-format:
+    type: int
```
While taking a deeper look I found that there is a common (?) definition in `display-controller.yaml`. Isn't it better to reuse this?
It uses int type as well.

jfischer-no replied:
TBH, I did not get it. You mean to add pixel format to display-controller.yaml?

kartben replied:
While I cannot say if this would be the right thing to do, I guess `rotation` and `inversion` might be in the same boat if we'd consider moving `pixel-format` to the common layer?

MrMarteng replied:
There already is a public header file and a yaml definition for the pixel-format configuration (`include/dt-bindings/display/panel.h` and `display-controller.yaml`). I would simply use that. This also makes the internal conversion obsolete, as these definitions are already in line with the lcd driver api.

jfischer-no replied:
I see, thanks for the point, but to me include/dt-bindings/display/panel.h looks a bit weird. Not sure why string enum could not be used. But that can be improved/fixed later. Btw, pixel-format should only be used to describe hardware, also "display-controller <-> panel-connection".

MrMarteng replied:
I used the existing pixel format definition from `lcd-controller.yaml` and `panel.h` for now. 
For orientation I used your approach with string enums, which could work commonly for all display drivers offering rotation/orientation feature.

For now the mixed usage of existing defines (pixel-format from panel.h) and the string enums for orientation is working fine with the benefit, that there is not conversion of configuration needed. The configured values match the display api defines.

I get your point but I think there needs to be a separate PR to clean up these common display parameters.

MrMarteng commented on include/zephyr/dt-bindings/display/gc9x01x.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_DISPLAY_GC9X01X_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_DISPLAY_GC9X01X_H_
+
+/* Pixel formats */
+#define GC9X01X_PIXEL_FORMAT_RGB565 0U
+#define GC9X01X_PIXEL_FORMAT_RGB888 1U
+
+/* Display rotation */
+#define GC9X01X_ORIENTATION_NORMAL 0U
+#define GC9X01X_ORIENTATION_90     90U
+#define GC9X01X_ORIENTATION_180    180U
+#define GC9X01X_ORIENTATION_270    270U
```
Can you point me to the comment you are referring to?
Or do you refer to the proposal to use a string like in pixel-format?

jfischer-no replied:
Yes, I mean can you use enum type to describe panel orientation, like here https://docs.zephyrproject.org/latest/build/dts/api/api.html#c.DT_ENUM_IDX

MrMarteng replied:
See comment above. Corrected in [83df655](https://github.com/zephyrproject-rtos/zephyr/commit/83df655b188b3a1ccdef966afd36d8f00d0e437d)

MrMarteng commented on drivers/display/display_gc9x01x.h:
```diff
@@ -0,0 +1,101 @@
+#include <zephyr/sys/util.h>
+
+/* Command registers */
+#define GC9X01X_CMD_SLPIN     (0x10U) /* Enter Sleep Mode */
```
fixed in [83df655](https://github.com/zephyrproject-rtos/zephyr/commit/83df655b188b3a1ccdef966afd36d8f00d0e437d)

MrMarteng commented on dts/bindings/display/galaxycore,gc9x01x.yaml:
```diff
@@ -0,0 +1,181 @@
+
+      The RESET pin of the GC9X01X is active low.
+      If connected directly the MCU pin should be configured
+      as active low.
```
Thank you.
Fixed in [83df655](https://github.com/zephyrproject-rtos/zephyr/commit/83df655b188b3a1ccdef966afd36d8f00d0e437d)

MrMarteng commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,738 @@
+
+	int retval;
+	struct spi_buf tx_buf = {.buf = &cmd, .len = 1U};
+	struct spi_buf_set tx_bufs = {.buffers = &tx_buf, .count = 1U};
```
corrected at all occurrences in [83df655](https://github.com/zephyrproject-rtos/zephyr/commit/83df655b188b3a1ccdef966afd36d8f00d0e437d)

fabiobaltieri replied:
I think this still has to be addressed, `int ret`? no other display driver use retval, saves few line wrappings

MrMarteng commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,738 @@
+	struct spi_buf_set tx_bufs = {.buffers = &tx_buf, .count = 1U};
+
+	retval = gpio_pin_set_dt(&config->cmd_data, GC9X01X_GPIO_LEVEL_CMD);
+	if (retval < 0) {
```
That's a good note. In the mentation case I think that's ok as there is not much that can be done when setting the gpio is failing. But I found another location in `gc9x01x_regs_init` where a proper evaluation was required ([83df655](https://github.com/zephyrproject-rtos/zephyr/commit/83df655b188b3a1ccdef966afd36d8f00d0e437d))

MrMarteng commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,738 @@
+	 * any manufacturing defects.
+	 * This is to allow time for the supply voltages and clock circuits stabilize
+	 */
+	k_msleep(GC9X01X_SLEEP_IN_OUT_DURATION_MS + 30U);
```
I am used to consequently use the U specified when the constant needs to be unsigned. So I would leave that for consistency.

MrMarteng replied:
@jfischer-no Could you please have a look. I think your comment is still pending.

MrMarteng replied:
@jfischer-no I reviewed your comment again and got you now. Fixed the definition in [56d8ba0](https://github.com/zephyrproject-rtos/zephyr/commit/56d8ba0ff2977428459b0159b634c622459851f8)

MrMarteng commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,738 @@
+		pixel_format = PIXEL_FORMAT_RGB_888;
+	}
+
+	/* Set pixel format. */
```
fixed in [83df655](https://github.com/zephyrproject-rtos/zephyr/commit/83df655b188b3a1ccdef966afd36d8f00d0e437d). See also comment above on device tree configuration value mapping. 
Custom header file is no longer needed.

faxe1008 reviewed (dismissed):
LGTM :+1:

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,691 @@
+	int retval;
+
+	/* Enable inter-command mode */
+	retval = gc9x01x_transmit(dev, GC9X01X_CMD_INREGEN1, NULL, 0);
+	retval |= gc9x01x_transmit(dev, GC9X01X_CMD_INREGEN2, NULL, 0);
```
this seems to somewhat imply that it may be useful to fail an spi transaction and go ahead with the others, but the reality is that if any one of these fails, something real bad happened, and there's no point delaying an error and trying to do everything else

how about bailing at the first fail? also use `ret` and save of the crazy indentation down there

```c
ret = gc9x01x_transmit(dev, GC9X01X_CMD_INREGEN1, NULL, 0);
if (ret < 0) {
        return ret;
}
```

MrMarteng replied:
You are absolutely right. Fixed in [a1256a9](https://github.com/zephyrproject-rtos/zephyr/commit/a1256a9156de7aca7ae9d46e4335310a8bb595bb)

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,691 @@
+	/* Enable Tearing line */
+	retval |= gc9x01x_transmit(dev, GC9X01X_CMD_TEON, NULL, 0);
+
+	if (retval == 0) {
+		return 0;
+	} else {
+		return -EIO;
+	}
```
see my comment above, return on the individual errors, `return 0;` here

MrMarteng replied:
Fixed above [a1256a9](https://github.com/zephyrproject-rtos/zephyr/commit/a1256a9156de7aca7ae9d46e4335310a8bb595bb)

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,691 @@
+	return 0;
+}
+
+static int gc9x01x_read(const struct device *dev, const uint16_t x, const uint16_t y,
+			const struct display_buffer_descriptor *desc, void *buf)
+{
+	LOG_ERR("Reading not supported");
+	return -ENOTSUP;
+}
+
+static void *gc9x01x_get_framebuffer(const struct device *dev)
+{
+	LOG_ERR("Direct framebuffer access not supported");
+	return NULL;
+}
+
+static int gc9x01x_set_contrast(const struct device *dev, const uint8_t contrast)
+{
+	LOG_ERR("Set contrast not supported");
+	return -ENOTSUP;
+}
```
the stubs in `include/zephyr/drivers/display.h` already check for some non implemented functions, I think you can drop all of these

MrMarteng replied:
Fixed in [a1256a9](https://github.com/zephyrproject-rtos/zephyr/commit/a1256a9156de7aca7ae9d46e4335310a8bb595bb)

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,691 @@
+#ifdef CONFIG_PM_DEVICE
+static int gc9x01x_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	int retval = 0;
```
no need to initialize this

MrMarteng replied:
Fixed in [a1256a9](https://github.com/zephyrproject-rtos/zephyr/commit/a1256a9156de7aca7ae9d46e4335310a8bb595bb)

fabiobaltieri commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,664 @@
+	if (retval < 0) {
+		return retval;
+	}
+	gc9x01x_transmit(dev, GC9X01X_CMD_INREGEN2, NULL, 0);
```
Sorry what I mean is to check it individually after every write and return if any fail. It'll be a bit spammy on the code but it's just for the init, I think it's well worth it. It is somewhat redundant but there's some cases where it may be useful, for example I know that the Nordic SPI in DMA mode have a limit on the transfer size, any one of those may hit it while the other don't so it's nice to have it exit at the right spot.

faxe1008 commented on drivers/display/display_gc9x01x.c:
```diff
@@ -0,0 +1,701 @@
+	}
+
+	/* Apply generic configuration */
+	retval =
```
Maybe make this `ret` so that the indent is not needed?

MrMarteng replied:
I didn't like that as well. Thanks for your porposal. Now the retail variable names are no longer consistent but I prefer readability as well.
FIxed in [5342707](https://github.com/zephyrproject-rtos/zephyr/commit/5342707641120c7b96c5ad63d62eb22b159f3b1b)

fabiobaltieri replied:
> Now the retail variable names are no longer consistent

I think that we may be all suggesting to use `ret` in the whole file.

MrMarteng replied:
OMG! It looks like I am meanwhile code blind and just react on comments.
Fixed in [c045f5c](https://github.com/zephyrproject-rtos/zephyr/commit/c045f5ce0a5f4aa4fb68a0bd7804e7944cc6564b)

amrithvenkat commented on dts/bindings/display/galaxycore,gc9x01x.yaml:
```diff
@@ -0,0 +1,157 @@
+  This driver implements support for various GC9xxx graphics
+  controllers and different display sizes. It has been validated
+  for following controllers:
+   - GC91A01A: (Waveshare 240x240, 1.28inch round lcd display 240x240)
```
```suggestion
   - GC9A01A: (Waveshare 240x240, 1.28inch round lcd display 240x240)
```

MrMarteng replied:
OMG! It looks like I am meanwhile code blind and just react on comments.
Fixed in [c045f5c](https://github.com/zephyrproject-rtos/zephyr/commit/c045f5ce0a5f4aa4fb68a0bd7804e7944cc6564b)

fabiobaltieri commented on dts/bindings/display/galaxycore,gc9x01x.yaml:
```diff
@@ -0,0 +1,158 @@
+
+  Here is an example to define a display interface:
+
+  / {
+    pwmleds {
+      compatible = "pwm-leds";
+      backlight_pwm: backlight_pwm {
+        pwms = <&ledc0 0 PWM_MSEC(10) PWM_POLARITY_NORMAL>;
+      };
+    };
+  };
```
one more: this is now outdated

MrMarteng replied:
Thought we want to leave this in for reference.
Fixed in [60c20cb](https://github.com/zephyrproject-rtos/zephyr/commit/60c20cb0e4f999ebe39a5e27b1f7fae6c01a5df4)

kartben commented on drivers/display/Kconfig.gc9x01x:
```diff
@@ -0,0 +1,11 @@
+	default y
+	depends on DT_HAS_GALAXYCORE_GC9X01X_ENABLED
+	select SPI
+	select PWM
```
drop?

MrMarteng replied:
Fixed in [810703f](https://github.com/zephyrproject-rtos/zephyr/commit/810703f196aaa5242daf43eb983aeca14d0fb18b)

kartben commented on tests/drivers/build_all/display/app.overlay:
```diff
@@ -23,6 +23,13 @@
 			status = "okay";
 		};
 
+		test_pwm1: pwm@55551111 {
+			compatible = "vnd,pwm";
+			#pwm-cells = <3>;
+			reg = < 0x55551111 0x1000 >;
+			status = "okay";
+		};
+
```
drop?

MrMarteng replied:
Thank you @kartben

MrMarteng replied:
Fixed in [810703f](https://github.com/zephyrproject-rtos/zephyr/commit/810703f196aaa5242daf43eb983aeca14d0fb18b)

kartben reviewed (changes requested):
sorry for having reviewed in chunks -- I have spotted what might be a few more issues or inconsistencies. Thanks!

MrMarteng commented on dts/bindings/display/galaxycore,gc9x01x.yaml:
```diff
@@ -0,0 +1,149 @@
+    type: string
+    default: "normal"
+    enum:
+      - "normal"
```
Using string enum type was a request from @jfischer-no .
The befit of using this approach is that missconfiguration can be detected at compile-time

jfischer-no replied:
I agree with @kartben , "0" would be more consistent and is a string like "90".

MrMarteng commented on drivers/display/CMakeLists.txt:
```diff
@@ -21,6 +21,7 @@ zephyr_library_sources_ifdef(CONFIG_ST7789V		display_st7789v.c)
 zephyr_library_sources_ifdef(CONFIG_ST7735R		display_st7735r.c)
 zephyr_library_sources_ifdef(CONFIG_STM32_LTDC		display_stm32_ltdc.c)
 zephyr_library_sources_ifdef(CONFIG_RM68200		display_rm68200.c)
+zephyr_library_sources_ifdef(CONFIG_GC9X01X		display_gc9x01x.c)
```
Aligned order in CMake file and Config. [090e023](https://github.com/zephyrproject-rtos/zephyr/commit/090e023674adc12091bf1000ffd6f82054354df5)

MrMarteng commented on dts/bindings/display/galaxycore,gc9x01x.yaml:
```diff
@@ -0,0 +1,149 @@
+description: |
+  GC9X01X display driver.
+
+  This driver implements support for various GC9xxx graphics
```
fixed in [090e023](https://github.com/zephyrproject-rtos/zephyr/commit/090e023674adc12091bf1000ffd6f82054354df5)

jfischer-no reviewed (approved):
Thanks. It would be nice to get https://github.com/zephyrproject-rtos/zephyr/pull/67012#discussion_r1459017862 resolved.

joelguittet commented:
Amazing job done here, thanks @MrMarteng and to all the reviewers :-)

MrMarteng commented:
> Amazing job done here, thanks @MrMarteng and to all the reviewers :-)

Thank you for all your feedback and support on that PR!!

joelguittet commented:
I'm just looking at the doc to update my own board and I see 2 errors:
- There is a wrong reference to `GC9X01X_PIXEL_FORMAT_RGB565` in the device tree sample. `pixel-format` is set twice! This one is minor.
- There is `#include <zephyr/dt-bindings/display/gc9x01x.h>` at https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_gc9x01x.c#L11. This does not exist anymore, I suppose this need to be `panel.h` instead. Can you confirm ?

I'm actually updating https://github.com/zephyrproject-rtos/zephyr/pull/67311, I can insert a commit to fix it if you have no time. Let me know if required. Else you will need a PR.

It's strange it has passed the Github Actions here by the way ....

MrMarteng commented:
> I'm just looking at the doc to update my own board and I see 2 errors:
> 
> - There is a wrong reference to `GC9X01X_PIXEL_FORMAT_RGB565` in the device tree sample. `pixel-format` is set twice! This one is minor.
> 
> - There is `#include <zephyr/dt-bindings/display/gc9x01x.h>` at https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_gc9x01x.c#L11. This does not exist anymore, I suppose this need to be `panel.h` instead. Can you confirm ?
> 
> 
> 
> I'm actually updating https://github.com/zephyrproject-rtos/zephyr/pull/67311, I can insert a commit to fix it if you have no time. Let me know if required. Else you will need a PR.
> 
> 
> 
> It's strange it has passed the Github Actions here by the way ....

I'll have a look latest tomorrow morning. I was continuasly checking proper functionality with my board. Give me a few hours to double check this. I will create a hotfix short term if required.

joelguittet commented:
Good thanks !

MrMarteng commented:
> Good thanks !

I did a clean checkout and can confirm that the driver isn't building properly. Sorry for that.
Fix will follow tomorrow - it's too late today

joelguittet commented:
@MrMarteng for your information, tested on my side: https://github.com/zephyrproject-rtos/zephyr/pull/67311/commits/18eedc4c00798d227926b389d78daf925e58accd and https://github.com/zephyrproject-rtos/zephyr/pull/67311/commits/4fd2ef8a517aed97c6764510a903b05fa611b461

ljunquera commented:
Do you know in which release this will be included?

joelguittet commented:
@ljunquera the driver is included in master and is part of the next release 3.6.0 that should land very soon (rc2 published).
